### ISSUE TITLE
ViewportTexture does not work in 3.0

### ISSUE THREAD

**Issue by curly-brace** [godotengine/godot#7998]
dunno if it is known or not, but ViewportTextures just don't work when launching the preview, but they are seen in editor window tho.

What i do - create viewport, add some controlls to it, check VFlip (dunno if it is bug too). Add Quad, add here fixed material and set ViewportTexture as its albedo texture.

I see everything in editor window, but only material color is seen on previewing. Normal textures work fine tho.

**Comment by curly-brace** [comment]
here is sample project https://www.dropbox.com/s/ujohhfsuotlr8hg/vp_test_30.zip?dl=0

**Comment by nunodonato** [comment]
I could replicate this (using master)

**Comment by RodZill4** [comment]
I couldn't get the ViewportTexture to work (only tested under Windows) in the first alpha. The texture is shown only in the editor, and when it has been created. When running the scene or when reloading the project, the ViewportTexture will be black.

See the "render_to_texture" scene in this project (you will have to re-create a viewport texture for the cube material emission texture and start the animation player in the editor to see it work):
[godot3_test.zip](https://github.com/godotengine/godot/files/1182041/godot3_test.zip)

It seems the ViewportTexture fails to find the viewport in the scene

**Comment by BastiaanOlij** [comment]
I'll try and find some time this weekend to play around with this, hopefully its something simple in the order in which things are constructed.

**Comment by BastiaanOlij** [comment]
I'm afraid I'm not sure how this is supposed to work and why it fails however the problem starts here:

https://github.com/godotengine/godot/blob/master/scene/main/viewport.cpp#L86

get_local_scene() does not return a value when the scene is loaded and as a result the viewport isn't being bound to the viewport texture. Loading your example project actually results in the material not working properly. The only way I can get it to work is by clearing the viewport_path in the editor and resetting it. It seems get_local_scene only behaves properly in this scenario. 

Digging a little deeper it looks like the function that returns the local scene is only set within the editor and is NULL in all other scenarios. 

@reduz I'm afraid this needs your wisdom.

**Comment by RodZill4** [comment]
The other scenario where the local scene would be non-null is where PackedScene uses the Resource::duplicate_for_local_scene method. Not sure when this applies (I tried to export the project but the ViewportTexture still does not show correctly), and I'm not familiar enough with Godot to dig that deep in the code. ;)
This Resource::get_local_scene() method is only used in ViewportTexture, so I guess it's new.

**Comment by NathanWarden** [comment]
I just ran into this same issue. I can record a video showing the issue if this would help :)

**Comment by NathanWarden** [comment]
Here's a video of the issue:

https://youtu.be/M3oFyPPqmHY

**Comment by hi-ogawa** [comment]
I'm not so certain but I think I found the reason.
The case where `ViewportTexture` works is when `ViewportTexture` is referenced from `Node` directly, which is, for example, there is `Sprite -> ViewporTexutre` in the scene.
But, the example shown in this issue is the scene which has the following structure:

```
MeshInstance (Node) -> SpatialMaterial (Resource) -> ViewportTexutre (Resource)
```

Here `ViewportTexutre` is only referenced from another `Resource` in a nested way and not from any `Node`.

It seems this tricky referencing is handled around the lines here https://github.com/godotengine/godot/blob/master/scene/resources/packed_scene.cpp#L224, and currently nested Resource is not considered, I assume.

But, that being said, actually part of me says, the concept of "local to scene" Resource doesn't fit the godot's Node-Resource paradigm and this feels like an architectural issue of ViewportTexture.

**Comment by reduz** [comment]
Actually this is not a bug and more by design. If you have:
`
MeshInstance (Node) -> SpatialMaterial (Resource) -> ViewportTexutre (Resource)`

You have to also set the SpatialMaterial as local to scene. for what you are intending to do to work.
I think the editor should probably be more obvious or user friendly when you do this. Will leave open as a reminder to make this more user friendly and documented.

**Comment by reduz** [comment]
maybe in 3.1

**Comment by NathanWarden** [comment]
@hi-ogawa Using a Sprite3D worked, thanks! :)

@reduz Should this be able to work on a material that has "Local To Scene" enabled? I tried this using a material directly on the node in the Material foldout, in the GeometryInstance foldout, and on the local Mesh resource itself (with "Local To Scene" also checked on the Mesh). None of those worked, except using a Sprite3D. Is it a bug if it isn't working on a material? It seems like it is since I can see someone wanting to use this on a curved CCTV or something like that.

**Comment by thismarty** [comment]
Just curious as to why this is 'closed' while the 'bug' is still present in Godot 3.0 today?

**Comment by thismarty** [comment]
Just curious, is this in the crosshairs?  Will we be able to attach viewport textures to materials directly without the sprit futz in 3.1?

**Comment by BastiaanOlij** [comment]
@thismarty viewports have a new property, get_texture if I'm not mistaken, you'll need to write a line of code but you can attach the viewport texture directly to your material

**Comment by NathanWarden** [comment]
@BastiaanOlij Thanks for mentioning this as I didn't know about it and sounds like a perfectly reasonable solution... will have to give it a try :)

**Comment by thismarty** [comment]
@BastiaanOlij Thanks for the tip but I'm not sure I follow you.  Do I attach a script to the Viewport calling get_texture(my-viewport-texture)?  Still feels like a futz though, don't you think?  Hopefully in 3.1 it'll 'just work'.   ;-)

**Comment by BastiaanOlij** [comment]
@thismarty like this:
https://github.com/BastiaanOlij/godot-sponza/blob/godot-sponza-vr/scripts/settings_vr.gd#L98

Note btw that I was just checking out the gui_in_3d demo source code that used to make use of this trick but they have now been changed to using proper viewport textures so I have a feeling things may have already been fixed so I would check that demo out and see how it works.

One thing I remember that was important that someone mentioned was to make sure "local to scene" is ticked on these nodes.

**Comment by BastiaanOlij** [comment]
@curly-brace I'll close this issue, as far as I'm aware this is all fixed in the latest master and mostly fixed in 3.0.2/3.0.3 by the introduction of the proxy. I've been happily using viewport textures all over the place in many of my recent projects. Feel free to re-open it if you feel there is more that should be done here.

**Comment by GammaGames** [comment]
To anyone else finding this issue in v3.0.6, check the viewport path. When I would use the built-in picker it would set the viewport path to something like `/root/EditorNode/@@5/@@6/.../Viewport`, when it only had to be the path in the tree without the base.
